crates/matrix-sdk-appservice/src/lib.rs

```diff
@@ -627,6 +632,89 @@ impl AppService {
         webserver::warp_filter(self.clone())
     }
 
+    /// Receive an incoming [transaction], pushing the contained events to
+    /// active virtual clients.
+    ///
+    /// [transaction]: https://spec.matrix.org/v1.2/application-service-api/#put_matrixappv1transactionstxnid
+    pub async fn receive_transaction(
+        &self,
+        transaction: push_events::v1::IncomingRequest,
+    ) -> Result<()> {
+        let client = self.get_cached_client(None)?;
+
+        // Find membership events affecting members in our namespace, and update
+        // membership accordingly
+        for event in transaction.events.iter() {
+            let event = match event.deserialize() {
+                Ok(AnyRoomEvent::State(AnyStateEvent::RoomMember(event))) => event,
+                _ => continue,
+            };
+            if !self.user_id_is_in_namespace(event.state_key())? {
+                continue;
+            }
+            let localpart = event.state_key().localpart();
+            client
```

Comment:
We'll probably want to store these as a single transaction instead of saving each member separately. A TODO + open issue will do.

Reply:
Is that possible when I don't have access to the private methods in the `matrix-sdk` crate?

Reply:
It's not possible currently, there's now an issue where we can add what else we would like from the store: https://github.com/matrix-org/matrix-rust-sdk/issues/730.